engine/server/engine/server/enclave_rest_api_handler.go

```
@@ -0,0 +1,857 @@
+	apiContainerGrpcPort := enclaveAPIContainerInfo.GrpcPortInsideEnclave
+	apiContainerIP := enclaveAPIContainerInfo.BridgeIpAddress
+
+	enclaveAPIContainerHostMachineInfo := enclaveInfo.ApiContainerHostMachineInfo
+	if connectOnHostMachine && enclaveAPIContainerHostMachineInfo == nil {
+		logrus.Infof("No API container info is available for enclave %s", enclaveInfo.EnclaveUuid)
+		return nil, nil
+	}
+	if connectOnHostMachine && (enclaveAPIContainerHostMachineInfo != nil) {
+		apiContainerGrpcPort = enclaveAPIContainerHostMachineInfo.GrpcPortOnHostMachine
+		apiContainerIP = enclaveAPIContainerHostMachineInfo.IpOnHostMachine
+	}
+
+	grpcServerAddress := fmt.Sprintf("%v:%v", apiContainerIP, apiContainerGrpcPort)
+	grpcConnection, err := grpc.Dial(grpcServerAddress, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, stacktrace.Propagate(err, "Expected to be able to create a GRPC client connection on address '%v', but a non-nil error was returned", grpcServerAddress)
+	}
+	return grpcConnection, nil
+}
+
+func (manager enclaveRuntime) GetApiClientOrResponseError(enclave_uuid string) (*rpc_api.ApiContainerServiceClient, *api_type.ResponseInfo) {
+	client, err := manager.GetGrpcClientForEnclaveUUID(enclave_uuid)
+	if err != nil {
+		return nil, &api_type.ResponseInfo{
+			Type:    api_type.ERROR,
+			Message: "Couldn't retrieve connection with enclave",
+			Code:    http.StatusInternalServerError,
+		}
+	}
+	if client == nil {
+		return nil, &api_type.ResponseInfo{
+			Type:    api_type.INFO,
+			Message: fmt.Sprintf("enclave '%s' not found", enclave_uuid),
+			Code:    http.StatusNotFound,
+		}
+	}
+	return client, nil
+}
+
+func (manager enclaveRuntime) GetGrpcClientForEnclaveUUID(enclave_uuid string) (*rpc_api.ApiContainerServiceClient, error) {
+	err := manager.refreshEnclaveConnections()
+	if err != nil {
+		return nil, err
+	}
+
+	client, found := manager.remoteApiContainerClient[enclave_uuid]
+	if !found {
+		return nil, nil
```

I think it should return an error here or the caller should check for nil values

---

Yes, and the client [does check for nil values](https://github.com/kurtosis-tech/kurtosis/blob/07c7f8e47f0dfab5a5fabea5b64b8cde1bd21ec7/engine/server/engine/server/enclave_rest_api_handler.go#L719)